pkg/location/type_git.go

```diff
@@ -0,0 +1,123 @@
+package location
+
+import (
+	"fmt"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/GoogleContainerTools/kpt/internal/errors"
+	"github.com/GoogleContainerTools/kpt/internal/util/parse"
+)
+
+type Git struct {
+	// Repo is the git repository the package.
+	// e.g. 'https://github.com/kubernetes/examples.git'
+	Repo string
+
+	// Directory is the sub directory of the git repository.
+	// e.g. 'staging/cockroachdb'
+	Directory string
+
+	// Ref can be a Git branch, tag, or a commit SHA-1.
+	Ref string
+}
+
+var _ Reference = Git{}
+
+type GitLock struct {
+	Git
+
+	// Commit is the SHA-1 for the last fetch of the package.
+	// This is set by kpt for bookkeeping purposes.
+	Commit string
+}
+
+var _ ReferenceLock = GitLock{}
+
+func NewGit(location string, opts ...Option) (Git, error) {
+	return newGit(location, makeOptions(opts...))
+}
+
+func newGit(location string, opt options) (Git, error) {
+	// args[1] is "" for commands that do not require an output path
+	gitTarget, err := parse.GitParseArgs(opt.ctx, []string{location, ""})
+	var zero parse.GitTarget
+	if err != nil || gitTarget == zero {
+		return Git{}, err
+	}
+
+	dir := gitTarget.Directory
+	if strings.HasPrefix(dir, "/") {
+		dir, err = filepath.Rel("/", gitTarget.Directory)
+		if err != nil {
+			return Git{}, err
+		}
+	}
+
+	return Git{
+		Repo:      gitTarget.Repo,
+		Directory: dir,
+		Ref:       gitTarget.Ref,
+	}, nil
+}
+
+func parseGit(location string, opt options) (Reference, error) {
```

Comment:
this may be unneeded (since newGit parses?)

Reply:
It's a subtle difference - but the different return value means parseGit can be used as a parser type - but the newGit call can be returned directly from NewGit without assert the type back to the struct.

TBH, parseGit could also be moved to an inline func in the option. That could be cleaner....